expression/util_test.go

```diff
@@ -320,6 +322,51 @@ func (s *testUtilSuite) TestFilterOutInPlace(c *check.C) {
 	c.Assert(filtered[0].(*ScalarFunction).FuncName.L, check.Equals, "or")
 }
 
+func (s *testUtilSuite) TestHashGroupKey(c *check.C) {
+	ctx := mock.NewContext()
+	sc := &stmtctx.StatementContext{TimeZone: time.Local}
+	eTypes := []types.EvalType{types.ETInt, types.ETReal, types.ETDecimal, types.ETString, types.ETTimestamp, types.ETDatetime, types.ETDuration}
+	tNames := []string{"int", "real", "decimal", "string", "timestamp", "datetime", "duration"}
+	for i := 0; i < len(tNames); i++ {
+		ft := eType2FieldType(eTypes[i])
+		if eTypes[i] == types.ETDecimal {
+			ft.Flen = 0
```

Comment:
What's this for?

Reply:
In the ```HashGroupKey``` function, when encoding the decimal type, the ```ft.len``` can't be less than 0. See the implementation in [EncodeDecimal](https://github.com/pingcap/tidb/blob/master/util/codec/decimal.go#L28) and [here](https://github.com/pingcap/tidb/blob/master/types/mydecimal.go#L1160) for more details.